### PR TITLE
Implement Specht modules for diagrams

### LINKED ISSUE #34883: Implement Specht modules for arbitrary diagrams
<div id="comment:0"></div>

We implement Specht modules arbitrary diagrams as a submodule of the symmetric group algebra.

We also include a method to just compute the dimension to avoid having to (potentially) fully echelonize the basis for constructing the submodule.

CC:  @fchapoton @anneschilling @zabrocki @saliola @darijgr

Component: **combinatorics**

Keywords: **Specht module**

Author: **Travis Scrimshaw**

_Issue created by migration from https://trac.sagemath.org/ticket/34883_

### PR DESCRIPTION
**Pull request by tscrim**
Fixes #34883.

### CODE REVIEW

**Comment by tscrim** [comment]
@zabrocki @saliola @darijgr Here is the PR for the branch that was previously there.

File: src/sage/combinat/diagram.py
Comment (fchapoton) [inline src/sage/combinat/diagram.py]: should be base_ring

File: src/sage/combinat/permutation.py
Comment (fchapoton) [inline src/sage/combinat/permutation.py:1246]: why having both `__mul__` and `_mul_` ?

**Comment by fchapoton** [comment]
I would prefer to add just one method everywhere, not the second method about rank.

File: src/sage/combinat/permutation.py
Comment (tscrim) [reply src/sage/combinat/permutation.py:1246]: Same reason we have them for other elements: When we know the object is in the correct parent, we can avoid the call through the coercion framework. Plus we need `_mul_()` for `prod`.

**Comment by tscrim** [comment]
> I would prefer to add just one method everywhere, not the second method about rank.

The rank computation is faster as it doesn't need to do the full echelonization. Since it can often be the point of interest (at least with more arbitrary shapes from what I've heard), I wanted to have easy access to it, plus the Specht module itself.

File: src/sage/combinat/partition.py
Comment (fchapoton) [inline src/sage/combinat/partition.py:5499]: maybe factor out a function `specht_module_builder` that takes a ring, a size and cells ? Instead of doing these double imports everywhere ?

**Comment by tscrim** [comment]
Okay, something is seriously borked up with how GH is syncing as I have merged in develop, which is what is leading to the massive file differences. @mkoeppe @dimpase What is going on with this?

**Comment by dimpase** [comment]
I don't have anything to do with releases - `develop` is updated by @vbraun

**Comment by dimpase** [comment]
my advice is to rebase, as far as it is easy. Leads to much clearer branches.

File: src/sage/combinat/partition.py
Comment (tscrim) [reply src/sage/combinat/partition.py:5499]: I don't really see the benefit. To me, it adds more indirection and complexity to save on just doing one import at each point of the code instead of two (as the second import is still done). To me, it is equivalent to having
```python
class Bar:
    def __init__(self, x):
        self._x = x

def foo(x):
    from above import Bar
    return Bar(x)
```
which I would want to just call `Bar(x)` within my code.

**Comment by tscrim** [comment]
@dimpase I am of the opinion rewriting history is bad (and it can mean I have to rehandle all merge conflicts; although that isn't the case here IIRC), but I will try. However, you are part of the GH transition team, and this is an issue with developing Sage with GH.

**Comment by tscrim** [comment]
@dimpase Actually, I do have merge conflicts to deal with...

**Comment by dimpase** [comment]
There were a number of very large edits done recently, e.g. replacing explicit references to trac.sagemath.org all over the sources.

This will not happen often.

**Comment by tscrim** [comment]
The rebase seems to have fixed the GH issue, but there should have been no difference with `develop`.

**Comment by mkoeppe** [comment]
@tscrim I didn't see what GitHub was showing to the diff; but I've just looked at your branch before the rebase (cca882762fa41c40b3ce5e688ad754d3c37f3ef0). 
The first parent of cca882762fa41c40b3ce5e688ad754d3c37f3ef0 is 10.0.beta3 – so you merged a branch *into* the development version. (Compare `git log cca882762fa41c40` and `git log --first-parent cca882762fa41c40`.) This is not good practice, and I can imagine that it may confuse tools that show diffs regarding what is to be considered the upstream branch. Instead, go to the branch and merge the `develop` branch into it.

**Comment by tscrim** [comment]
@mkoeppe It was showing 524 files changed. I also _really_ never want to rollback my version of Sage to a previous version just to merge in `develop`. This needlessly wastes a huge amount of time with recompiling stuff that didn't actually change (beyond the timestamps)., Not to mention it is far more commands and a good diff tool should be comparing files rather than by commits (which seems to me like it would be very brittle).

**Comment by mkoeppe** [comment]
It's documented, by the way, here: https://doc.sagemath.org/html/en/developer/advanced_git.html#minimize-the-recompilation-time

**Review by fchapoton — APPROVED** [review]
ok, let it be

**Comment by tscrim** [comment]
Thank you!

**Comment by fchapoton** [comment]
in my opinion, it is not required to merge develop after getting positive review

**Comment by tscrim** [comment]
> in my opinion, it is not required to merge develop after getting positive review

Quite possibly; I am somewhat of the same opinion. Although it is one click on GH and makes sure there are no merge conflicts…

**Comment by dimpase** [comment]
By merging (even better, rebasing, if possible), you're alleviating the risk that the branch does not merge cleanly any more. 

This is something that ought to be improved in our development workflow. IMHO most everyone thinks that our `develop` branch is too slow with its updates.

**Comment by tscrim** [comment]
But not everyone; in particular, I am not of that opinion. Although having a more all-PRs-currently-merged branch that doesn’t change would be nice to have. Yet that might end up being more work for Volker…

**Comment by dimpase** [comment]
my point is that with other GitHub projects I contribute to, once your PR is approved, the rest is  not my worry. Here, you still have to often watch your approved contribution to get all mouldy and often in need of more polish, because bitrot...

**Comment by mkoeppe** [comment]
> Although it is one click on GH and makes sure there are no merge conflicts…

GH checks for merge conflicts already when you view the PR; there's no need to push the merge button for that. When it displays "branch is out of date", that means it can be merged cleanly.

I think the standing guidance from @vbraun still applies: Don't touch a PR that has the "positive review" label. (It may already be merged into Volker's branch.) With the modification: If GH indicates that there is a merge conflict, then it makes sense to resolve the conflicts – it typically can't be merged in this case.

**Comment by tscrim** [comment]
@mkoeppe Do the tests run against the latest develop too?

@dimpase I think there is only so much that can be done for that because of a higher volume of changes that SageMath development has. The only thing I could think of to prevent that would be the every-accepted-PR-merged branch I mentioned before. Usually the biggest lag in beta versions comes around release time and just after, but I think that comes from Volker having to do too much.

**Comment by dimpase** [comment]
> Do the tests run against the latest develop too?

as far as I know, they are not running with an explicit trigger such as a `git push`.
So in that sense, updating the branch triggers the tests against the new base branch.

**Comment by dimpase** [comment]
> every-accepted-PR-merged branch

as far as I know, that's what large Python orgs, such as cPython, scipy, etc., use.

**Comment by vbraun** [comment]
Merge conflict

**Comment by tscrim** [comment]
Trivial rebase.

**Pull request merged**
